### ISSUE TITLE
Signals not cleared by fork

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11061]
Migrated from [rt.perl.org#82580](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=82580) (status was 'resolved')

Searchable as RT82580$

**Comment by p5pRT** [comment]
### From philipgwyn@gmail.com

#### Created by philipgwyn@gmail.com



Perl\_csighandler\(\) \(the sigaction\(\) signal handler\) sets a flag in
PL\_psig\_pend\[\]\.  These flags are checked in Perl\_despatch\_signals\(\)
via PERL\_ASYNC\_CHECK in Perl\_runops\_standard\(\)\, if I undertand the code
correctly\.

This introduces a race condition for pp\_fork\(\)&#8203;:
&nbsp;    \- PERL\_ASYNC\_CHECK 
&nbsp;        All flags clear
&nbsp;    \- A signal arrives
&nbsp;        A flag is set
&nbsp;    \- pp\_fork\(\) is called
&nbsp;    \- fork\(\) is called
&nbsp;        The flag is set in both the child and the parent\.

The quick fix&#8203;:
&nbsp;    \- pp\_fork\(\) masks all signals
&nbsp;    \- fork\(\) is called
&nbsp;    \- in the parent&#8203;:
&nbsp;        \- unmask signals
&nbsp;        \- return PID
&nbsp;    \- in the child&#8203;:
&nbsp;        \- clear signal flags
&nbsp;        \- unmask signals

There is a problem with this though&#8203;: Unless I mis\-understand how signal
masking works\, signals that arrive between mask\(\) call land the unmask\(\)
call \(in the parent\) will be lost\.

The hard fix \(the shared\-nothing model\)&#8203;:
&nbsp;    \- Create a one\-way pipe
&nbsp;    \- Signal handlers write one byte \(the signal \#\) to the pipe
&nbsp;    \- PERL\_ASYNC\_CHECK does a non\-blocking read\, gets the byte\(s\)
&nbsp;    \- The signal is dispatched via Perl\_sighandler as normal

Included is a short script that exercises the problem\.  This script has 
provoked the problem with perl versions 5\.8\.8\, 5\.10\.1\, 5\.12\.1\, 5\.12\.2 \(and
probably more\) on Linux \(CentOS\, Gentoo\, Ubuntu\) and Mac OS X\.

\#\!/usr/bin/perl

use strict;
use warnings;
use Carp;

use POSIX "&#8203;:sys\_wait\_h";

my $parent  = $$;
my $todo    = 16;       \# Maximum number of child processes in play\.
my $running = 1;

print "my pid=$$ parent pid=$parent \- this is the master control process\, the only forker\\n";

$SIG\{CHLD\} = sub \{
&nbsp;    confess "my pid=$$ parent pid=$parent \- child process mistakenly got SIGCHLD somehow\\n" if $parent \!= $$;

&nbsp;    \# NB \- waitpid\(\) \< 0 is legal on MSWin32\.
&nbsp;    while \(waitpid\(\-1\, WNOHANG\) > 0\) \{
&nbsp;        $todo\+\+;
&nbsp;        $running = 0 if $?;
&nbsp;    \}
\};

sub forker \{
&nbsp;    confess "my pid=$$ expected=$parent \- child process somehow got into the main loop\\n" if $$ \!= $parent;

&nbsp;    while \($todo > 0\) \{
&nbsp;        my $child\_pid = fork\(\);
&nbsp;        confess "my pid=$$ parent pid=$parent \- fork failed&#8203;: $\!" unless
&nbsp;        defined $child\_pid;

&nbsp;        \# Child here\. Trigger SIGCHLD in the parent\.
&nbsp;        exit unless $child\_pid;

&nbsp;        \# Parent\.  One fewer child to fork\.
&nbsp;        $todo\-\-;
&nbsp;    \}
\}

\# Known that this spins\.
\# We want to stress fork\(\) and $SIG\{CHLD\}\.
forker\(\) while $running;

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.12.2:

Configured by fil at Fri Jan 21 13:53:48 EST 2011.

Summary of my perl5 (revision 5 version 12 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.18-194.17.1.el5, archname=x86_64-linux
    uname='linux george-dev5.localdomain 2.6.18-194.17.1.el5 #1 smp wed sep 29 12:50:31 edt 2010 x86_64 x86_64 x86_64 gnulinux '
    config_args='-de -Dprefix=/home/fil/perl5/perlbrew/perls/perl-5.12.2'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.1.2 20080704 (Red Hat 4.1.2-48)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64 /usr/local/lib64
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.12.2:
    /home/fil/perl5lib/lib64/perl5/site_perl/5.8.8
    /home/fil/perl5lib/lib/perl5/site_perl/5.8.8
    /home/fil/perl5/perlbrew/perls/perl-5.12.2/lib/site_perl/5.12.2/x86_64-linux
    /home/fil/perl5/perlbrew/perls/perl-5.12.2/lib/site_perl/5.12.2
    /home/fil/perl5/perlbrew/perls/perl-5.12.2/lib/5.12.2/x86_64-linux
    /home/fil/perl5/perlbrew/perls/perl-5.12.2/lib/5.12.2
    .


Environment for perl 5.12.2:
    HOME=/home/fil
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/fil/perl5/perlbrew/bin:/home/fil/perl5/perlbrew/perls/perl-5.12.2/bin:/usr/kerberos/bin:/usr/local/bin:/bin:/usr/bin:/home/fil/bin
    PERL5LIB=/home/fil/perl5lib/lib64/perl5/site_perl/5.8.8:/home/fil/perl5lib/lib/perl5/site_perl/5.8.8
    PERLBREW_PATH=/home/fil/perl5/perlbrew/bin:/home/fil/perl5/perlbrew/perls/perl-5.12.2/bin
    PERLBREW_PERL=perl-5.12.2
    PERLBREW_ROOT=/home/fil/perl5/perlbrew
    PERLBREW_VERSION=0.15
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From philipgwyn@gmail.com

Here is a test case \(from Rocco Caputo\) that triggers the problem using
USR1&#8203;:

\#\!/usr/bin/perl

use strict;
use warnings;
use Carp;

use POSIX "&#8203;:sys\_wait\_h";

my $parent  = $$;
my $todo    = 16;        \# Maximum number of child processes in play\.
my $running = 1;

print "my pid=$$ parent pid=$parent \- this is the master control
process\, the only forker\\n";

$SIG\{USR1\} = sub \{
&nbsp;    return if $$ == $parent;
&nbsp;    confess "my pid=$$ parent pid=$parent \- child process got SIGUSR1
destined for parent\\n" if $parent \!= $$;
\};

$SIG\{CHLD\} = sub \{

&nbsp;    \# NB \- waitpid\(\) \< 0 is legal on MSWin32\.
&nbsp;    $todo\+\+ while waitpid\(\-1\, WNOHANG\) > 0;
\};

sub forker \{
&nbsp;    confess "my pid=$$ expected=$parent \- child process somehow got into
the main loop\\n" if $$ \!= $parent;

&nbsp;    while \($todo > 0\) \{
&nbsp;        my $child\_pid = fork\(\);
&nbsp;        confess "my pid=$$ parent pid=$parent \- fork failed&#8203;: $\!" unless
defined $child\_pid;

&nbsp;        \# Child here\. Trigger SIGCHLD in the parent\.
&nbsp;        unless \($child\_pid\) \{
&nbsp;            kill USR1 => $parent;
&nbsp;            exit;
&nbsp;        \}

&nbsp;        \# Parent\.  One fewer child to fork\.
&nbsp;        $todo\-\-;
&nbsp;    \}
\}

\# Known that this spins\.
\# We want to stress fork\(\) and $SIG\{CHLD\}\.
forker\(\) while $running;

**Comment by p5pRT** [comment]
philipgwyn@gmail.com - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From blgl@stacken.kth.se

2011\-01\-21 kl\. 22\.37 skrev Philip Gwyn&#8203;:
> There is a problem with this though&#8203;: Unless I mis\-understand how signal
> masking works\, signals that arrive between mask\(\) call land the unmask\(\)
> call \(in the parent\) will be lost\.

You misunderstand\.  It's ignored signals that are lost\, not blocked ones\.


/Bo Lindbergh

**Comment by p5pRT** [comment]
### From philipgwyn@gmail.com

On Sun Jan 23 04&#8203;:44&#8203;:57 2011\, blgl@&#8203;stacken\.kth\.se wrote&#8203;:
> You misunderstand\.  It's ignored signals that are lost\, not blocked ones\.

A short test shows that while CHLD makes it through\, waitpid\(\) will
always return \-1\.  Tested on 5\.8\.8\, 5\.10\.1\, 5\.12\.2\.

> Note that a return value of "\-1" could mean that child processes are 
> being automatically reaped\, as described in perlipc\.

**Comment by p5pRT** [comment]
### From philipgwyn@gmail.com

[fork-6](https://rt-archive.perl.org/perl5/Ticket/Attachment/794390/385154/fork-6)

**Comment by p5pRT** [comment]
### From philipgwyn@gmail.com

http&#8203;://search\.cpan\.org/dist/IPC\-SafeFork/

**Comment by p5pRT** [comment]
### From @Leont

On Tue Jan 25 12&#8203;:06&#8203;:00 2011\, Leolo wrote&#8203;:
> On Sun Jan 23 04&#8203;:44&#8203;:57 2011\, blgl@&#8203;stacken\.kth\.se wrote&#8203;:
> > You misunderstand\.  It's ignored signals that are lost\, not blocked
ones\.
> 
> A short test shows that while CHLD makes it through\, waitpid\(\) will
> always return \-1\.  Tested on 5\.8\.8\, 5\.10\.1\, 5\.12\.2\.

Perl reaps processes when closing filehandles created through «open
"\-|"»\, there's nothing weird or worrisome about your observation\.

Leon

**Comment by p5pRT** [comment]
### From @Leont

On Tue Apr 17 04&#8203;:09&#8203;:16 2012\, LeonT wrote&#8203;:
> On Tue Jan 25 12&#8203;:06&#8203;:00 2011\, Leolo wrote&#8203;:
> > On Sun Jan 23 04&#8203;:44&#8203;:57 2011\, blgl@&#8203;stacken\.kth\.se wrote&#8203;:
> > > You misunderstand\.  It's ignored signals that are lost\, not blocked
> ones\.
> > 
> > A short test shows that while CHLD makes it through\, waitpid\(\) will
> > always return \-1\.  Tested on 5\.8\.8\, 5\.10\.1\, 5\.12\.2\.
> 
> Perl reaps processes when closing filehandles created through �open
> "\-|"�\, there's nothing weird or worrisome about your observation\.
> 
> Leon

There's a patch that fixes this in the smoke\-me/leont/signals\-fork
branch\. Given it's a non\-deterministic issue\, I don't have a test though\.

Leon

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed in eb3d0a5826\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
